File: Elixir/lib/store.ex
defmodule Store do
  use GenServer

  defstruct content: %{}, get_counter: 0, set_counter: 0, del_counter: 0

  def start_link(d) do
    GenServer.start_link(__MODULE__, d, name: MyStore)
  end

  @impl true
  def init(_) do
    {:ok, %Store{}}
  end

  @impl true
  def handle_cast({:set, key, value}, %Store{content: con, get_counter: get_c, set_counter: set_c, del_counter: del_c}) do
    record = Record.new(key, value)
    new_map = Map.put(con, key, record)
    {:noreply, %Store{content: new_map, get_counter: get_c, set_counter: set_c + 1, del_counter: del_c}}
  end

  @impl true
  def handle_cast(_, state) do
    {:noreply, state}
  end

  @impl true
  def handle_call({:get, key}, _from, state) do
    content = Map.get(state, :content)
    record = Map.get(content, key)
    get_counter = Map.get(state, :get_counter)
    get_counter = get_counter + 1
    new_state = Map.put(state, :get_counter, get_counter)
    {:reply, record, new_state}
  end

  @impl true
  def handle_call({:del, key}, _from, state) do
    content = Map.get(state, :content)
    record = Map.get(content, key)
    new_content = Map.delete(content, key)
    del_counter = Map.get(state, :del_counter)
    new_state = Map.put(state, :content, new_content)
    new_state = Map.put(new_state, :del_counter, del_counter + 1)
    {:reply, record, new_state}
  end

  @impl true
  def handle_call(:del_counter, _from, state) do
    {:reply, Map.get(state, :del_counter), state}
  end

  @impl true
  def handle_call(:get_counter, _from, state) do
    {:reply, Map.get(state, :get_counter), state}
  end

  @impl true
  def handle_call(:set_counter, _from, state) do
    {:reply, Map.get(state, :set_counter), state}
  end

  @impl true
  def handle_call(:values, _from, state) do
    content = Map.get(state, :content)
    values = Map.values(content)
    {:reply, values, state}
  end

  @impl true
  def handle_call(_, _from, state) do
    {:reply, nil, state}
  end

  @spec set(bitstring, bitstring) :: :ok
  def set(key, value) do
    GenServer.cast(MyStore, {:set, key, value})
  end

  @spec get(bitstring) :: Record | nil
  def get(key) do
    GenServer.call(MyStore, {:get, key})
  end

  def del(key) do
    GenServer.call(MyStore, {:del, key})
  end

  def get_counter() do
    GenServer.call(MyStore, :get_counter)
  end

  def set_counter() do
    GenServer.call(MyStore, :set_counter)
  end

  def del_counter() do
    GenServer.call(MyStore, :del_counter)
  end

  def values() do
    GenServer.call(MyStore, :values)
  end
end

File: Python/store.py
import asyncio
from typing import Dict, Optional

from record import Record


class Store:
    content: Dict[str, Record]
    content_lock: asyncio.Lock
    get_counter: int
    get_counter_lock: asyncio.Lock
    set_counter: int
    set_counter_lock: asyncio.Lock
    del_counter: int
    del_counter_lock: asyncio.Lock

    def __init__(self):
        self.content = {}
        self.content_lock = asyncio.Lock()
        self.get_counter = 0
        self.get_counter_lock = asyncio.Lock()
        self.set_counter = 0
        self.set_counter_lock = asyncio.Lock()
        self.del_counter = 0
        self.del_counter_lock = asyncio.Lock()

    async def set(self, key: str, value: str) -> Optional[Record]:
        record = Record(key, value)
        await self.content_lock.acquire()
        ret = Record()
        ret.invalidate()
        if key in self.content:
            ret = self.content[key]
        self.content[key] = record
        self.content_lock.release()
        await self.set_counter_lock.acquire()
        self.set_counter += 1
        self.set_counter_lock.release()
        return ret

    async def get(self, key: str) -> Optional[Record]:
        await self.content_lock.acquire()
        ret = Record()
        ret.invalidate()
        if key in self.content:
            ret = self.content[key]
        self.content_lock.release()
        await self.get_counter_lock.acquire()
        self.get_counter += 1
        self.get_counter_lock.release()
        return ret

    async def delete(self, key: str) -> Optional[Record]:
        await self.content_lock.acquire()
        ret = Record()
        ret.invalidate()
        if key in self.content:
            ret = self.content[key]
        del self.content[key]
        self.content_lock.release()
        await self.del_counter_lock.acquire()
        self.del_counter += 1
        self.del_counter_lock.release()
        return ret

    async def set_counter_num(self) -> int:
        await self.set_counter_lock.acquire()
        ret = self.set_counter
        self.set_counter_lock.release()
        return ret

    async def get_counter_num(self):
        await self.get_counter_lock.acquire()
        ret = self.get_counter
        self.get_counter_lock.release()
        return ret

    async def del_counter_num(self):
        await self.del_counter_lock.acquire()
        ret = self.del_counter
        self.del_counter_lock.release()
        return ret
